velox/functions/prestosql/ArrayAllMatch.cpp

```diff
@@ -0,0 +1,136 @@
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    exec::LocalDecodedVector arrayDecoder(context, *args[0], rows);
+    auto& decodedArray = *arrayDecoder.get();
+
+    auto flatArray = flattenArray(rows, args[0], decodedArray);
+    auto offsets = flatArray->rawOffsets();
+    auto sizes = flatArray->rawSizes();
+
+    std::vector<VectorPtr> lambdaArgs = {flatArray->elements()};
+    auto numElements = flatArray->elements()->size();
+
+    SelectivityVector finalSelection;
+    if (!context.isFinalSelection()) {
+      finalSelection = toElementRows<ArrayVector>(
+          numElements, *context.finalSelection(), flatArray.get());
+    }
+
+    VectorPtr matchBits;
+    auto elementToTopLevelRows = getElementToTopLevelRows(
+        numElements, rows, flatArray.get(), context.pool());
+
+    // Loop over lambda functions and apply these to elements of the base array,
+    // in most cases there will be only one function and the loop will run once.
+    context.ensureWritable(rows, BOOLEAN(), result);
+    auto flatResult = result->asFlatVector<bool>();
+    exec::LocalDecodedVector bitsDecoder(context);
+    auto it = args[1]->asUnchecked<FunctionVector>()->iterator(&rows);
+
+    // Turn off the ThrowOnError flag and reset errors in the
+    // context to nullptr, so we only handle errors in this function itself.
+    auto topLevelThrowOnError = context.throwOnError();
+    ScopedVarSetter throwOnError(context.mutableThrowOnError(), false);
+    while (auto entry = it.next()) {
+      auto elementRows =
+          toElementRows<ArrayVector>(numElements, *entry.rows, flatArray.get());
+      auto wrapCapture = toWrapCapture<ArrayVector>(
+          numElements, entry.callable, *entry.rows, flatArray);
+      entry.callable->apply(
+          elementRows,
+          finalSelection,
+          wrapCapture,
+          &context,
+          lambdaArgs,
+          elementToTopLevelRows,
+          &matchBits);
+
+      bitsDecoder.get()->decode(*matchBits, elementRows);
+      entry.rows->applyToSelected([&](vector_size_t row) {
+        auto errors = context.errors();
+        auto size = sizes[row];
+        auto offset = offsets[row];
+        auto allMatch = true;
+        auto hasNull = false;
+
+        for (auto i = 0; i < size; ++i) {
+          auto idx = offset + i;
+          if (bitsDecoder->isNullAt(idx)) {
+            hasNull = true;
+          } else if (!bitsDecoder->valueAt<bool>(idx)) {
+            allMatch = false;
+            break;
+          }
+        }
+
+        // Errors for individual array elements should be suppressed only if the
+        // outcome can be decided by some other array element, e.g. if there is
+        // another element that returns 'false' for the predicate.
+        if (allMatch) {
+          if (topLevelThrowOnError) {
```

Comment:
I missed this last time. Functions should not throw, but rather report errors via context.setError().

https://facebookincubator.github.io/velox/develop/scalar-functions.html#try-expression-support

Reply:
Ok, I've adjust the logic according to the doc.

Reply:
Thanks.

I'm not sure this code is correct. Functions should unconditionally report errors via context.setError. However, we are explicitly disabling throwOnError in the context, hence, calling context.setError won't throw, but also since context errors will get restored to what they were before calling this function, context.setError below will get lost. 

```
          if (topLevelThrowOnError) {
            context.throwOnError(row);
          } else {
            context.setError(row, context.getError(row));
          }
```

I think we need to do something else. Specifically, we need to evaluate `entry.callable` and record errors for individual array elements. We then need to process these and we should not trust the value of the 'predicate' for elements that have an error.